gnomad/variant_qc/pipeline.py

```diff
@@ -320,3 +328,97 @@ def generate_sib_stats(
     ).rows()
 
     return sib_stats_ht
+
+
+def train_rf_model(
+    ht: hl.Table,
+    rf_features: List[str],
+    tp_expr: hl.expr.BooleanExpression,
+    fp_expr: hl.expr.BooleanExpression,
+    fp_to_tp: float = 1.0,
+    num_trees: int = 500,
+    max_depth: int = 5,
+    test_expr: Optional[
+        Union[hl.expr.BooleanExpression, List[hl.expr.IntervalExpression]]
+    ] = None,
+) -> Tuple[hl.Table, pyspark.ml.PipelineModel]:
+    """
+    Perform random forest (RF) training using a Table annotated with features and training data.
+
+    .. note::
+
+        This function uses `train_rf` and extends it by:
+            - Adding an option to apply the resulting model to test intervals which are withheld from training.
+            - An option to exclude variants from training.
+            - Uses a false positive (FP) to true positive (TP) ratio to determine what variants to use for RF training.
+
+    The returned Table includes the following annotations:
+        - rf_train: indicates if the variant was used for training of the RF model.
+        - rf_label: indicates if the variant is a TP or FP.
+        - rf_test: indicates if the variant was used in testing of the RF model.
+        - features: global annotation of the features used for the RF model.
+        - features_importance: global annotation of the importance of each feature in the model.
+        - test_results: results from testing the model on variants/intervals defined by `test_expr`.
+
+    :param ht: Table annotated with features for the RF model and the positive and negative training data.
+    :param rf_features: List of column names to use as features in the RF training.
+    :param tp_expr: TP training expression.
+    :param fp_expr: FP training expression.
+    :param fp_to_tp: Ratio of FPs to TPs for creating the RF model. If set to 0, all training examples are used.
+    :param num_trees: Number of trees in the RF model.
+    :param max_depth: Maxmimum tree depth in the RF model.
+    :param test_expr: An expression specifying variants or a list of intervals to hold out for testing and use for evaluation only.
+    :return: Table with TP and FP training sets used in the RF training and the resulting RF model.
+    """
+    if isinstance(test_expr, list):
+        rf_test_expr = hl.literal(test_expr).any(
+            lambda interval: interval.contains(ht.locus)
+        )
+    elif isinstance(test_expr, hl.expr.BooleanExpression):
+        rf_test_expr = test_expr
+    else:
+        rf_test_expr = False
+
+    ht = ht.annotate(tp=tp_expr, fp=fp_expr, rf_test=rf_test_expr,)
```

Comment:
I kind of like how you've used `_tp` and `_fp` in `sample_training_examples` as it gives a clue that these annotations won't be returned. Up to you, but I'd support the change here too :)